bootstore/src/schemes/v0/fsm.rs

```diff
@@ -0,0 +1,820 @@
+            return Err(ApiError::AlreadyInitialized);
+        };
+        let pkgs = create_pkgs(rack_uuid, initial_membership.clone())
+            .map_err(|e| ApiError::RackInitFailed(e))?;
+        let mut iter = pkgs.expose_secret().into_iter();
+        let our_pkg = iter.next().unwrap().clone();
+
+        // Move into an initialized state
+        self.state = State::InitialMember {
+            pkg: our_pkg,
+            distributed_shares: BTreeMap::new(),
+        };
+
+        let packages: BTreeMap<Baseboard, SharePkg> = initial_membership
+            .into_iter()
+            .filter(|peer| *peer != self.id)
+            .zip(iter.cloned())
+            .collect();
+
+        let _ = self.request_manager.new_init_rack_req(
+            now,
+            rack_uuid,
+            packages.clone(),
+            &self.connected_peers,
+        );
+
+        Ok(())
+    }
+
+    /// Are we still waiting for `InitAck` responses from peers?
+    pub fn is_rack_initializing(&self) -> bool {
+        self.request_manager.has_init_rack_req()
+    }
+
+    /// Initialize a node added after rack initialization
+    ///
+    /// Persistence is required after a successful call to `init_learner`
+    pub fn init_learner(&mut self, now: Instant) -> Result<(), ApiError> {
+        if let Some((_, err)) = &self.rack_init_error {
+            return Err(err.clone());
+        }
+        let State::Uninitialized = self.state else {
+            return Err(ApiError::AlreadyInitialized);
+        };
+
+        if let Some(to) = self.connected_peers.first() {
+            let _ = self.request_manager.new_learn_sent_req(now, to.clone());
+        }
+        self.state = State::Learning;
+        Ok(())
+    }
+
+    /// This call is triggered locally after RSS runs, in order to retrieve
+    /// the `RackSecret` so that it can be used as input key material. It
+    /// starts a a key share retrieval process so that the `RackSecret` can
+    /// be reconstructed.
+    pub fn load_rack_secret(&mut self, now: Instant) -> Result<Uuid, ApiError> {
+        if let Some((_, err)) = &self.rack_init_error {
+            return Err(err.clone());
+        }
+        let request_id = match &self.state {
+            State::Uninitialized => return Err(ApiError::NotInitialized),
+            State::Learning { .. } => return Err(ApiError::StillLearning),
+            State::InitialMember { pkg, .. } => {
+                self.request_manager.new_load_rack_secret_req(
+                    now,
+                    pkg.rack_uuid,
+                    pkg.threshold,
+                    &self.connected_peers,
+                )
+            }
+            State::Learned { pkg } => {
+                self.request_manager.new_load_rack_secret_req(
+                    now,
+                    pkg.rack_uuid,
+                    pkg.threshold,
+                    &self.connected_peers,
+                )
+            }
+        };
```

Comment:
Okay, so you _could_ reduce some duplication with
```rust
        let (rack_uuid, threshold) = match &self.state {
            State::Uninitialized => return Err(ApiError::NotInitialized),
            State::Learning { .. } => return Err(ApiError::StillLearning),
            State::InitialMember { pkg, .. } => (pkg.rack_uuid, pkg.threshold),
            State::Learned { pkg } => (pkg.rack_uuid, pkg.threshold),
        };

        let request_id = self.request_manager.new_load_rack_secret_req(
            now,
            rack_uuid,
            threshold,
            &self.connected_peers,
        );
```
(up to you if it's worth it)

Reply:
Yes! I am a fan of this type of dedupe. Thanks!